File: OfflineTranscription/Models/AppError.swift
import Foundation

enum AppError: LocalizedError {
    case microphonePermissionDenied
    case microphonePermissionRestricted
    case modelDownloadFailed(underlying: Error)
    case modelLoadFailed(underlying: Error)
    case transcriptionFailed(underlying: Error)
    case audioSessionSetupFailed(underlying: Error)
    case noModelSelected
    case modelNotReady
    case audioInterrupted
    case translationUnavailable
    case translationFailed(underlying: Error)
    case ttsFailed(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .microphonePermissionDenied:
            return "Microphone access was denied. Please enable it in Settings."
        case .microphonePermissionRestricted:
            return "Microphone access is restricted on this device."
        case .modelDownloadFailed(let error):
            return "Failed to download the model: \(error.localizedDescription)"
        case .modelLoadFailed(let error):
            return "Failed to load the model: \(error.localizedDescription)"
        case .transcriptionFailed(let error):
            return "Transcription failed: \(error.localizedDescription)"
        case .audioSessionSetupFailed(let error):
            return "Audio session setup failed: \(error.localizedDescription)"
        case .noModelSelected:
            return "No transcription model selected."
        case .modelNotReady:
            return "The transcription model is not ready yet."
        case .audioInterrupted:
            return "Recording was interrupted by another app or phone call."
        case .translationUnavailable:
            return "Native translation is unavailable. Requires iOS 18.0+ with the Translation framework."
        case .translationFailed(let error):
            return "Translation failed: \(error.localizedDescription)"
        case .ttsFailed(let error):
            return "Text-to-speech failed: \(error.localizedDescription)"
        }
    }
}

File: VoicePingIOSAndroidOfflineSpeechTranslationAndroid/app/src/main/kotlin/com/jima/offlinetranscription/model/AppError.kt
package com.voiceping.offlinetranscription.model

sealed class AppError(val message: String) {
    class MicrophonePermissionDenied : AppError("Microphone access was denied. Please enable it in Settings.")
    class SystemAudioCaptureUnsupported :
        AppError("System audio capture requires Android 10 (API 29) or later.")
    class SystemAudioCapturePermissionDenied :
        AppError("System audio capture is not enabled. Tap 'Enable System Capture' and accept the prompt.")
    class NoMicrophoneSignal : AppError(
        "No microphone signal detected. Check that the microphone is not muted or blocked."
    )
    class NetworkUnavailable : AppError("No internet connection. Connect to download the model.")
    class ModelDownloadFailed(cause: Throwable) : AppError("Failed to download the model: ${cause.localizedMessage}")
    class ModelLoadFailed(cause: Throwable) : AppError("Failed to load the model: ${cause.localizedMessage}")
    class TranscriptionFailed(cause: Throwable) : AppError("Transcription failed: ${cause.localizedMessage}")
    class TranslationUnavailable : AppError(
        "Native on-device translation is unavailable on this Android version/device."
    )
    class TranslationFailed(cause: Throwable) : AppError("Translation failed: ${cause.localizedMessage}")
    class TtsFailed(cause: Throwable) : AppError("Text-to-speech failed: ${cause.localizedMessage}")
    class NoModelSelected : AppError("No transcription model selected.")
    class ModelNotReady : AppError("The transcription model is not ready yet.")
    class InsufficientStorage(needed: String, available: String) :
        AppError("Not enough storage to download model (need $needed, have $available available).")
}
